### PR TITLE
disposableFiles: fix edge cases again

### PR DESCRIPTION
**Pull request by justinmk3**
### PROBLEM: 

Some tests implicitly depend on the global `ExtensionDisposableFiles.INSTANCE`, other tests clear it, but it is not correctly re-initialized.

###  SOLUTION:

Use `clearInstance()`, and rearrange its logic to avoid a race condition. Hoist `TestExtensionDisposableFiles` into `FakeExtensionContext` so it can use `clearInstance()`.

### ANALYSIS:

Previous attempt to fix this in https://github.com/aws/aws-toolkit-vscode/pull/1133 was incomplete, and only "worked" by accident. Problem resurfaces in `feature/debugconfig` because it has more tests that depend on `ExtensionDisposableFiles.INSTANCE`.

Rearrange `clearInstance()` so that

    ExtensionDisposableFiles.INSTANCE = undefined

is the first step. Because `del.sync()` does file IO, node may switch to another context while it waits on the file IO.

Fixes this error (among others):

    1) localLambdaRunner
      "before all" hook in "localLambdaRunner":
    Error: ExtensionDisposableFiles already initialized
     at Function.<anonymous> (/Volumes/workplace/aws-toolkit-vscode/src/shared/utilities/disposableFiles.ts:78:19)
     at Generator.next (<anonymous>)
     at /Volumes/workplace/aws-toolkit-vscode/dist/src/shared/utilities/disposableFiles.js:12:71
     at new Promise (<anonymous>)
     at /Volumes/workplace/aws-toolkit-vscode/dist/src/shared/utilities/disposableFiles.js:8:12
     at Function.initialize (/Volumes/workplace/aws-toolkit-vscode/dist/src/shared/utilities/disposableFiles.js:79:16)
     at Function.<anonymous> (/Volumes/workplace/aws-toolkit-vscode/src/test/fakeExtensionContext.ts:72:44)
     at Generator.next (<anonymous>)
     at /Volumes/workplace/aws-toolkit-vscode/dist/src/test/fakeExtensionContext.js:12:71
     at new Promise (<anonymous>)
     at /Volumes/workplace/aws-toolkit-vscode/dist/src/test/fakeExtensionContext.js:8:12
     at Function.getNew (/Volumes/workplace/aws-toolkit-vscode/dist/src/test/fakeExtensionContext.js:54:16)
     at /Volumes/workplace/aws-toolkit-vscode/src/test/shared/codelens/localLambdaRunner.test.ts:22:36
     at Generator.next (<anonymous>)
     at /Volumes/workplace/aws-toolkit-vscode/dist/src/test/shared/codelens/localLambdaRunner.test.js:12:71
     at new Promise (<anonymous>)
     at /Volumes/workplace/aws-toolkit-vscode/dist/src/test/shared/codelens/localLambdaRunner.test.js:8:12
     at Context.<anonymous> (/Volumes/workplace/aws-toolkit-vscode/src/test/shared/codelens/localLambdaRunner.test.ts:21:23)
     at processImmediate (internal/timers.js:439:21)
     at process.topLevelDomainCallback (domain.js:131:23)

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/test/fakeExtensionContext.ts [hunk 1]
```diff
@@ -54,14 +56,32 @@ export class FakeExtensionContext implements ExtensionContext {
     public static async getNew(): Promise<FakeExtensionContext> {
         const ctx = new FakeExtensionContext()
         try {
-            ExtensionDisposableFiles.getInstance().dispose()
+            TestExtensionDisposableFiles.clearInstance()
         } catch {
-            await ExtensionDisposableFiles.initialize(ctx)
+            // Ignore.
         }
+        await TestExtensionDisposableFiles.initialize(ctx)
```
Comment (justinmk3) [inline src/test/fakeExtensionContext.ts:63]: the `catch` in the old code was supposed to be a `finally`. Typescript does not (yet?) support `try...finally` without `catch`.

File: src/test/fakeExtensionContext.ts [hunk 2]
```diff
@@ -54,14 +56,32 @@ export class FakeExtensionContext implements ExtensionContext {
...
+export class TestExtensionDisposableFiles extends ExtensionDisposableFiles {
+    public static ORIGINAL_INSTANCE = ExtensionDisposableFiles.INSTANCE
+
+    public static clearInstance() {
+        const instance = ExtensionDisposableFiles.INSTANCE
+        ExtensionDisposableFiles.INSTANCE = undefined
+        if (instance) {
+            del.sync([instance.toolkitTempFolder], { force: true })
+            instance.dispose()
+        }
```
Comment (justinmk3) [inline src/test/fakeExtensionContext.ts:77]: Note that this logic was rearranged (mentioned in the commit message): `ExtensionDisposableFiles.INSTANCE=undefined` is done first, because 

1. `del.sync()` does file IO
2. the node scheduler looks for pending handlers to execute while waiting on IO
3. that cause a race: other async handlers may try to use `ExtensionDisposableFiles.getInstance()` which then throws an exception

File: src/shared/utilities/disposableFiles.ts [hunk 3]
```diff
@@ -26,6 +26,10 @@ export class DisposableFiles implements vscode.Disposable {
         return this
     }
 
+    public isDisposed(): boolean {
+        return !!this._disposed
```
Comment (awschristou) [inline src/shared/utilities/disposableFiles.ts:30]: Is `!!` necessary?

File: src/test/fakeExtensionContext.ts (hunk 2 above)
Comment (awschristou) [reply src/test/fakeExtensionContext.ts:77]: It may be worth adding some comments to the code around this (more visible than commit messages)

File: src/test/fakeExtensionContext.ts [hunk 4]
```diff
@@ -54,14 +56,32 @@ export class FakeExtensionContext implements ExtensionContext {
...
         } catch {
-            await ExtensionDisposableFiles.initialize(ctx)
+            // Ignore.
         }
+        await TestExtensionDisposableFiles.initialize(ctx)
         return ctx
     }
 }
 
+export class TestExtensionDisposableFiles extends ExtensionDisposableFiles {
```
Comment (awschristou) [inline src/test/fakeExtensionContext.ts:68]: What is the value in extending ExtensionDisposableFiles? This class seems like it could be a freestanding test utility class.

File: src/test/fakeExtensionContext.ts (hunk 1 above)
Comment (awschristou) [inline src/test/fakeExtensionContext.ts:63]: TestExtensionDisposableFiles doesn't appear to have a static initialize method, which was confusing when I initially read this change. Consider changing to ExtensionDisposableFiles.initialize(...)

File: src/test/fakeExtensionContext.ts [hunk 5]
```diff
@@ -3,7 +3,9 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { ExtensionContext, Memento } from 'vscode'
+import * as vscode from 'vscode'
+import * as del from 'del'
+import { Memento } from 'vscode'
```
Comment (bryceitoc9) [inline src/test/fakeExtensionContext.ts:8]: Why not `vscode.Memento`?

File: src/test/fakeExtensionContext.ts [hunk 6]
```diff
@@ -54,14 +56,32 @@ export class FakeExtensionContext implements ExtensionContext {
...
 }
 
+export class TestExtensionDisposableFiles extends ExtensionDisposableFiles {
+    public static ORIGINAL_INSTANCE = ExtensionDisposableFiles.INSTANCE
+
+    public static clearInstance() {
+        const instance = ExtensionDisposableFiles.INSTANCE
+        ExtensionDisposableFiles.INSTANCE = undefined
+        if (instance) {
+            del.sync([instance.toolkitTempFolder], { force: true })
```
Comment (bryceitoc9) [inline src/test/fakeExtensionContext.ts:75]: Is this not covered by `instance.dispose()`?

File: src/shared/utilities/disposableFiles.ts (hunk 3 above)
Comment (justinmk3) [reply src/shared/utilities/disposableFiles.ts:30]: to convert to an explicit `boolean` ?

File: src/test/fakeExtensionContext.ts (hunk 4 above)
Comment (justinmk3) [reply src/test/fakeExtensionContext.ts:68]: one reason is that `INSTANCE` is `protected`

File: src/test/fakeExtensionContext.ts (hunk 6 above)
Comment (justinmk3) [reply src/test/fakeExtensionContext.ts:75]: I wondered that too but assumed that @kantc had a reason to do it this way

File: src/shared/utilities/disposableFiles.ts (hunk 3 above)
Comment (awschristou) [reply src/shared/utilities/disposableFiles.ts:30]: I don't understand, the declaration is ` private _disposed: boolean = false `

File: src/test/fakeExtensionContext.ts (hunk 4 above)
Comment (awschristou) [reply src/test/fakeExtensionContext.ts:68]: Ok, I overlooked that

File: src/shared/utilities/disposableFiles.ts (hunk 3 above)
Comment (justinmk3) [reply src/shared/utilities/disposableFiles.ts:30]: oh haha

File: src/test/fakeExtensionContext.ts (hunk 1 above)
Comment (justinmk3) [reply src/test/fakeExtensionContext.ts:63]: but if TestExtensionDisposableFiles later adds initialize(), then the caller must be updated.

Comment (awschristou) [reply src/test/fakeExtensionContext.ts:62]: Sounds like a problem to solve if/when that happens

Comment (justinmk3) [reply src/test/fakeExtensionContext.ts:62]: Doesn't that apply to all uses of subclasses?
